src/Init/Data/BitVec/Lemmas.lean

```diff
@@ -5127,6 +5128,84 @@ theorem msb_replicate {n w : Nat} {x : BitVec w} :
   simp only [BitVec.msb, getMsbD_replicate, Nat.zero_mod]
   cases n <;> cases w <;> simp
 
+/-! ### Vectors of bitvectors -/
+
+@[simp]
+theorem pop_splitLE : (splitLE m (n + 1) x).pop = splitLE m n (x.extractLsb' m _) := by
+  ext k hk i hi
+  simp at hk
+  have : m * (n - k - 1) + i < m * n := by
+    apply Nat.lt_of_lt_of_le (Nat.add_lt_add_left hi _)
+    rw [← Nat.mul_add_one]
+    apply Nat.mul_le_mul_left
+    omega
+  simp only [Nat.add_one_sub_one, splitLE, Vector.range, Vector.map_mk, Vector.pop_mk,
+    Vector.getElem_mk, Array.getElem_pop, Array.getElem_map, Array.getElem_range,
+    getElem_extractLsb', this, getLsbD_eq_getElem, ← Nat.add_assoc]
+  rw [Nat.sub_add_comm (Nat.le_of_lt hk), Nat.sub_add_comm (Nat.le_sub_of_add_le' hk),
+    Nat.mul_add m (n - k - 1), Nat.add_comm m]
+  simp
+
+@[simp]
+theorem back_splitLE : (splitLE m (n + 1) x).back = x.extractLsb' 0 m := by
+  ext i hi
+  simp [splitLE, Vector.back, Vector.range, Nat.add_sub_cancel_left]
+
+@[simp]
+theorem flattenBitVecLE_splitLE {m n : Nat} {x : BitVec (m * n)} :
+    (x.splitLE m n).flattenBitVecLE m = x := by
+  induction n with
+  | zero =>
+    simp [splitLE, Vector.flattenBitVecLE, Vector.flatMapBitVecLE, of_length_zero]
+  | succ n ih =>
+    simp only [Vector.flattenBitVecLE, Vector.flatMapBitVecLE, Nat.add_one_sub_one, pop_splitLE,
+      cast_eq, back_splitLE, id_eq] at ih ⊢
+    rw [ih]
+    ext i hi
+    by_cases hi' : i < m
+    · simp [getElem_append, hi', BitVec.getLsbD_eq_getElem hi]
+    · simp [getElem_append, hi', Nat.add_sub_of_le (Nat.le_of_not_lt hi'),
+        BitVec.getLsbD_eq_getElem hi]
+
+theorem getElem_flattenBitVecLE {m n : Nat} [NeZero m] (v : Vector (BitVec m) n) (i : Nat)
+    (hi : i < m * n) :
+    (v.flattenBitVecLE m)[i] =
+      (v[n - (i / m + 1)]'(Nat.sub_lt (n.pos_of_ne_zero fun hn =>
+        by subst hn; omega) (i / m).zero_lt_succ))[i % m]'(Nat.mod_lt _ m.pos_of_neZero) := by
+  induction n generalizing i with
+  | zero =>
+    simp at hi
+  | succ n ih =>
+    simp only [Vector.flattenBitVecLE, Vector.flatMapBitVecLE, Nat.add_one_sub_one, cast_eq, id_eq]
+    rw [Nat.mul_add, Nat.mul_one] at hi
+    apply (BitVec.getElem_append hi).trans
+    by_cases hi' : i < m
+    · simp [hi', Vector.back, Nat.mod_eq_of_lt hi', Nat.div_eq_of_lt hi']
+    · simp only [Vector.flattenBitVecLE] at ih
+      simp only [hi', ↓reduceDIte, ih v.pop (i - m) (by omega)]
+      simp [Vector.pop, ← Nat.mod_eq_sub_mod (Nat.ge_of_not_lt hi'),
+        ← Nat.div_eq_sub_div m.pos_of_neZero (Nat.ge_of_not_lt hi')]
+
+theorem getElem_splitLE {i : Nat} (hi : i < n) (x : BitVec (m * n)) :
+    (x.splitLE m n)[i] = x.extractLsb' (m * (n - i - 1)) m := by
+  simp [splitLE, Vector.range]
+
+@[simp]
+theorem splitLE_flattenBitVecLE {m n : Nat} [NeZero m] {v : Vector (BitVec m) n} :
+    (v.flattenBitVecLE m).splitLE m n = v := by
```

Comment:
```suggestion
@[simp]
theorem splitLE_flattenBitVecLE {m n : Nat} [NeZero m] {v : Vector (BitVec m) n} :
    (v.flattenBitVecLE m).splitLE m n = v := by
```

For the same reason as above, I believe we don't need the `[NeZero m]` constraint here, unless I am missing something?